Scale nodes by file size

zimonitrome opened the pull request:
## Summary

This PR adds optional file-node scaling by file size via `-S` / `--scale-by-file-size`.

When `-S` is not set, existing Gource behavior is unchanged.

Related issues: #91 #54 #147 #223

[Screencast from 2026-02-10 17:22:23_trimmed.webm](https://github.com/user-attachments/assets/a1b321bb-3e7f-447b-ab94-d93be495f81e)

## What changed

- Added file-size-based node scaling (`-S` / `--scale-by-file-size`).
- Added smooth interpolation when file size changes (so node size transitions are animated).
- Added a dedicated packing solver for scaled nodes to reduce wobble and avoid persistent overlap in dense directories.
- Added deterministic small-cluster shaping to prevent line/chain layouts for small groups of files.
- Added optional hover text for file size (`--show-file-size-on-hover`).

## Git size lookup changes

- Reworked Git size lookup to index blob sizes up front using:
  - `git cat-file --batch-all-objects --batch-check='%(objectname) %(objecttype) %(objectsize)'`
- During parsing, file sizes are looked up from this in-memory index (no per-event subprocess calls).
- Git raw log command now uses `--abbrev=40` so blob hashes match the prebuilt index.
- Existing log file compatibility:
  - If scaling is OFF, older/partial raw lines continue to parse.
  - If scaling is ON and required hash metadata is missing, parsing fails with a descriptive error.

## Custom log format

Custom log supports optional `file_size` as a 6th field:

`timestamp|username|type|file|colour|file_size`

`file_size` is used for `A`/`M` actions.

## CLI/options added

Primary feature flag:
- `-S`, `--scale-by-file-size`

Advanced tuning options:
- `--file-scale`
- `--dir-spacing`
- `--file-gravity`
- `--file-repulsion`
- `--show-file-size-on-hover`

Defaults are tuned so `-S` works reasonably without extra tuning.

## Notes

Current scaling uses a logarithmic mapping for readability across large size ranges.

zimonitrome committed:
Minor cleanup.

zimonitrome committed:
Add node scaling by file size.

acaudwell commented:
Cool looking feature. Thanks for the detailed overview of the changes.

acaudwell commented on src/gource_settings.cpp:
```diff
@@ -230,6 +238,8 @@ GourceSettings::GourceSettings() {
     arg_aliases["b"] = "background-colour";
     arg_aliases["c"] = "time-scale";
+    arg_aliases["S"] = "scale-by-file-size";
+    arg_aliases["S"] = "scale-by-file-size";
```
duplicate?

dagguh commented on src/dirnode.cpp:
```diff
@@ -573,7 +579,17 @@ float RDirNode::getArea() const{
+    if (gGourceSettings.scale_by_file_size) {
+        for (RFile* f : files) {
+            if (!f->isHidden()) {
+                float r = f->getSize() / 2.0f;
```
So this sets the radius to file size. It means the actual dot size will be proportional to file size squared. The area of the dots **will not be proportional** to file size.

acaudwell commented on src/formats/git.cpp:
```diff
@@ -127,7 +128,7 @@ std::string GitCommitLog::logCommand() {
 }
 
-GitCommitLog::GitCommitLog(const std::string& logfile) : RCommitLog(logfile, 'u') {
+GitCommitLog::GitCommitLog(const std::string& logfile) : RCommitLog(logfile, 'u'), m_repository_path(logfile) {
```
Maybe instead RCommitLog could keep a copy of logfile and then we can refer to it later with self.logfile

zimonitrome replied:
Implemented. `RCommitLog` now stores the original logfile path and exposes it via `getLogfilePath()`, and Git uses that context when needed.

acaudwell commented on src/formats/git.cpp:
```diff
@@ -215,30 +216,63 @@ bool GitCommitLog::parseCommit(RCommit& commit) {
             //this isnt a commit we are parsing, abort
             if(commit.timestamp == 0) return false;
 
+            if(!logf->getNextLine(line)) return false;
```
The logfile could be an existing file generated with a previous version of Gource's log file command without the hash. It should still continue to work with the existing logfile, so if this line is a file line (starting with ':'), and they didn't request scaling by file size, it needs to recover from that. If they are using file scaling and there's no hash it could fail to parse at that point, ideally with a descriptive error.

zimonitrome replied:
Implemented. Parser now tolerates legacy/partial raw lines when scaling is off, and emits descriptive errors when scaling is on but required blob-hash metadata is missing.

acaudwell commented on src/formats/git.cpp:
```diff
@@ -215,30 +216,63 @@ bool GitCommitLog::parseCommit(RCommit& commit) {
+
+                if (gGourceSettings.scale_by_file_size && status != "D") {
+                    char cmd_buff[2048];
+                    int written = snprintf(cmd_buff, 2048, "git --git-dir=%s/.git --work-tree=%s cat-file -s %s", m_repository_path.c_str(), m_repository_path.c_str(), dst_blob.c_str());
```
This seems to cause a performance issue doing this here as it blocks the UI while fetching the blob. Also this wont work if logfile is a file and not a directory. You can also see it block when you move the mouse cursor of the interactive timeline as it peeks at the part of the log under the cursor to get the timestamp.

I think for performance instead what would be good is we get all the file blob sizes up front right after we generate the git log, puts them into a data structure, and then we can look up into it here.

zimonitrome replied:
Implemented. I removed per-event `git cat-file` calls and replaced them with a one-time blob-size index build right after log generation, then O(1) lookups during parsing. Should avoid timeline-hover/UI stalls... I think. Maybe needs more testing.

acaudwell commented on src/file.cpp:
```diff
@@ -73,6 +77,24 @@ RFile::RFile(const std::string & name, const vec3 & colour, const vec2 & pos, in
 }
 
+void RFile::setFileSize(unsigned int new_file_size) {
+    file_size = new_file_size;
```
Would be nice to animate the file size changing, so it interpolates between the old file size and the new file size, maybe over the same period as the action laser beam. So it would keep the old size but also store the new size, and then it could use another variable like file_size_change_time_elapsed to work out the current size.

zimonitrome replied:
Implemented. File size changes now interpolate between old and new values over a short duration, so modifies animate instead of snapping.

zimonitrome commented:
I updated the branch to make the "physics layout" more stable and took in some some review feedback.

Main improvements since earlier revisions:

- Replaced per-file blob size subprocess calls with a one-time Git blob index build (`git cat-file --batch-all-objects ...`) and in-memory lookups.
- Added compatibility behavior for existing/older raw log files:
  - no scaling => continue parsing
  - scaling enabled but missing blob metadata => descriptive error
- Added `--abbrev=40` to the Git raw log command so blob hashes consistently match the index.
- Added file size transition animation (interpolated node size changes).
- Reworked scaled-node directory packing to reduce wobble/overlap in large clusters and avoid chain-like minima in smaller clusters.

Updated my OP to reflect the changes, also added a new video. Some behavior should maybe be changed? As in distance between directories (clusters)?

Might do a bit more testing but it seems pretty nice all in all right now!

zimonitrome commented:
* Added better Windows (+mingw) support
* Format file sizes in human readable format (i.e. KB, MB, GB, instead of just "bytes")
* Added scaling interpolation argument (float) where
    * 0.0: uniform (same as without any file scaling)
    * 1.0: log scaling (default
    * 2.0: linear scaling

Here I ran it with: 
```
gource ../myproject --scale-by-file-size --show-file-size-on-hover --dir-spacing 0.7 -s 1 -a 0.5 --hide filenames --file-size-scale-interpolation 1.5
```
<img width="1024" height="768" alt="image" src="https://github.com/user-attachments/assets/940b6620-4ef4-4800-95ec-ef8f0d69183e" />
